components/ai_chat/core/browser/model_service.cc

```diff
@@ -651,6 +652,43 @@ void ModelService::DeleteCustomModel(uint32_t index) {
   }
 }
 
+void ModelService::MaybeDeleteCustomModels(CustomModelPredicate predicate) {
+  ScopedListPrefUpdate update(pref_service_, kCustomModelsList);
+  std::vector<std::string> removed_keys;
+
+  // Remove models matching predicate (iterate in reverse to avoid index issues)
+  for (size_t i = update->size(); i > 0; --i) {
+    const base::Value::Dict& model_dict = (*update)[i - 1].GetDict();
+
+    if (predicate.Run(model_dict)) {
+      std::string removed_key = *model_dict.FindString(kCustomModelItemKey);
+      removed_keys.push_back(removed_key);
+
+      // Check if this is the default model
+      if (GetDefaultModelKey() == removed_key) {
+        pref_service_->ClearPref(kDefaultModelKey);
+        DVLOG(1) << "Default model key " << removed_key
+                 << " was removed. Cleared default model key.";
+        for (auto& obs : observers_) {
+          obs.OnDefaultModelChanged(removed_key, GetDefaultModelKey());
+        }
+      }
+
+      update->erase(update->begin() + (i - 1));
+    }
+  }
+
+  if (!removed_keys.empty()) {
+    InitModels();
+
+    for (const auto& removed_key : removed_keys) {
+      for (auto& obs : observers_) {
+        obs.OnModelRemoved(removed_key);
+      }
+    }
+  }
```

Comment:
why can't we notify the observer when we iterate? And then only need a bool to call `InitModels()` when any one of  the models is removed.